### PR TITLE
Safe Mode

### PR DESCRIPTION
**Pull request by hensha256**

### CODE REVIEW

File: src/entities/protocols/uniswap.ts [hunk 1]
```diff
@@ -56,10 +58,7 @@ export class UniswapTrade implements Command {
     // If the input currency is the native currency, we need to wrap it with the router as the recipient
     if (this.trade.inputAmount.currency.isNative) {
       // TODO: optimize if only one v2 pool we can directly send this to the pool
-      planner.addCommand(CommandType.WRAP_ETH, [
-        ROUTER_AS_RECIPIENT,
-        this.trade.maximumAmountIn(this.options.slippageTolerance).quotient.toString(),
-      ])
+      planner.addCommand(CommandType.WRAP_ETH, [ROUTER_AS_RECIPIENT, CONTRACT_BALANCE])
```
Comment (ewilz) [inline src/entities/protocols/uniswap.ts:61]: What if there's a UNiswap trade that takes WETH, then an NFT trade that takes ETH? Would this break that?

**Comment by ewilz** [comment]
change feels hefty, the odds of losing an ERC20 are so slim. The most minimally invasive change to the sdk IMO is to add a bool parameter to  `SwapRouterConfig` called `safetyMode` and add a sweepETH command at the end before encoding the "plan". Basically a 2 line change.

If we must be this thorough this looks fiiiiiine i guess 😏 but wouldnt want the sdk to get too out of hand with trying to every last unlikely case.

**Review by ewilz — APPROVED** [review]

File: src/entities/protocols/uniswap.ts [hunk 2]
```diff
@@ -53,6 +68,11 @@ export class UniswapTrade implements Command {
   encode(planner: RoutePlanner, _config: TradeConfig): void {
     let payerIsUser = true
 
+    let safetyMode = this.options.safetyMode ?? SafetyModes.Normal
+    let inputETHSweep = safetyMode == SafetyModes.InputOutputSweeps || safetyMode == SafetyModes.AllSweeps
```
Comment (ewilz) [inline src/entities/protocols/uniswap.ts:72]: maybe we rename to `shouldSweepETH` and `shouldSweepIntermediateTokens` for readability, know it's a bool flag

**Comment by ewilz** [comment]
per slack convo: could we enforce a certain slippage instead of sweeping erc20 tokens on intermediate swaps??

**Review by ewilz — CHANGES_REQUESTED** [review]
slippage in lieau of sweeping interediate??

File: src/entities/protocols/uniswap.ts [hunk 3]
```diff
@@ -29,9 +29,11 @@ export type FlatFeeOptions = {
 
 // the existing router permit object doesn't include enough data for permit2
 // so we extend swap options with the permit2 permit
+// when safe mode is enabled, the SDK will add extra token sweeps for security
```
Comment (dannydaniil) [inline src/entities/protocols/uniswap.ts:32]: should we change this to "when tokensToSweep"?

Comment (hensha256) [reply src/entities/protocols/uniswap.ts:32]: oh yeah good catch ty

**Pull request closed without merging**